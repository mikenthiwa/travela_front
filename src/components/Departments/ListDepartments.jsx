import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import LoadingDepartmentsTable from './DepartmentsTable';
import TemplatesPagination from '../ReminderSetup/TemplatesPagination';

class ListDepartments extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const { getAllDepartments, location: { search } } = this.props;
    getAllDepartments(search);
  }

  onPageChange = (direction) => {
    const {
      department: {pagination: {currentPage}},
      getAllDepartments,
      location: {search}
    } = this.props;
    const requestedPage = direction === 'next' ? currentPage + 1 : currentPage - 1;
    const query = `${search}${search ? '&' : '?'}page=${requestedPage}`;
    getAllDepartments(query);
  };

  Pagination = (currentPage, pageCount, onPageChange) => {
    return (
      <TemplatesPagination
        currentPage={currentPage ? currentPage : 1}
        pageCount={pageCount ? pageCount : 1}
        onPageChange={onPageChange} />
    );
  };

  render(){
    const {
      shouldOpen,
      closeModal,
      modalType,
      openModal,
      editDepartments,
      department: {departments, pagination, isLoadingDepartment}
    } = this.props;
    const { currentPage,  pageCount } = pagination;
    return (
      <Fragment>
        <LoadingDepartmentsTable
          editDepartments={editDepartments}
          departments={departments}
          shouldOpen={shouldOpen}
          openModal={openModal}
          closeModal={closeModal}
          modalType={modalType}
        />
        {!isLoadingDepartment && pageCount > 0 && this.Pagination(currentPage, pageCount, this.onPageChange)}
      </Fragment>
    );
  }
}

ListDepartments.propTypes = {
  getAllDepartments: PropTypes.func,
  editDepartments: PropTypes.func,
  shouldOpen: PropTypes.bool,
  closeModal: PropTypes.func,
  modalType: PropTypes.string,
  openModal: PropTypes.func,
  department: PropTypes.object,
  location: PropTypes.object,
};

ListDepartments.defaultProps = {
  getAllDepartments: null,
  editDepartments: null,
  shouldOpen: false,
  closeModal: null,
  modalType: '',
  openModal: null,
  department: {},
  location: {}
};

export default ListDepartments;
