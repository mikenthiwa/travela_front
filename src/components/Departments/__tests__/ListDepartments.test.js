import React from 'react';
import { Provider } from 'react-redux';
import sinon from 'sinon';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme/build';
import ListDepartments  from '../ListDepartments';

describe('<ListDepartments />', () => {
  const props = {
    getAllDepartments: jest.fn(),
    editDepartment: jest.fn(),
    addDepartment: jest.fn(),
    openModal: sinon.spy(() => Promise.resolve()),
    closeModal: jest.fn(),
    shouldOpen: false,
    modalType: '',
    department: {
      singleDepartment: {
        id: 1,
        name: 'Operations',
        parentDepartment: 1,
        parentDepartments: {
          name: 'Andela',
        },
        creator: {
          fullName: 'Adeniyi Adeyokunnu',
        },
        createdBy: 1
      },
      isLoadingDepartment: true,
      pagination: {
        pageCount: 1,
        currentPage: 1,
        totalCount: 1
      },
      departments: [
        {
          id: 1,
          name: 'Operations',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 2,
          name: 'Technology',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 3,
          name: 'Success',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 4,
          name: 'Facilities',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 5,
          name: 'Partner Engineering',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 6,
          name: 'TDD',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 7,
          name: 'Apprenticeship',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 8,
          name: 'Finance',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 9,
          name: 'Travel',
          parentDepartment: null,
          createdBy: 1
        },
        {
          id: 10,
          name: 'Marketing',
          parentDepartment: null,
          createdBy: 1
        },{
          id: 11,
          name: 'Communications',
          parentDepartment: null,
          createdBy: 1
        },
      ]
    },
    location: {},
    history: {
      push: jest.fn()
    },
    modal: {
      modal: {
        modal: {
          openModal: jest.fn(),
          closeModal: jest.fn(),
          shouldOpen: false
        }
      }
    }
  };

  const state = {};
  const mockStore = configureStore();
  const store = mockStore (state);
  let wrapper;

  beforeEach(() => {
    wrapper = mount( <ListDepartments {...props} store={store} />);
  });

  it('renders appropriately', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should list departments with Pagination changing to next', () => {
    wrapper.setProps({ department: {
      ...props.department,
      isLoadingDepartment: false,
      pagination: {
        pageCount: 2,
        currentPage: 1,
        totalCount: 11
      }
    }});

    const inst = wrapper.instance();

    inst.onPageChange('next');
    inst.onPageChange = jest.fn();

    inst.onPageChange('next');
    wrapper.find('.pagination Button .pagination__button').at(0).simulate('click');
    expect(inst.onPageChange).toHaveBeenCalled();
  });

  it('should list departments with Pagination changing to previous', () => {
    wrapper.setProps({
      location: {
        search: true,
      },
      department: {
        ...props.department,
        isLoadingDepartment: false,
        pagination: {
          pageCount: 2,
          currentPage: 2,
          totalCount: 11
        }
      }
    });

    const inst = wrapper.instance();

    inst.onPageChange('previous');
    inst.onPageChange = jest.fn();

    inst.onPageChange('previous');
    wrapper.find('.pagination Button .pagination__button').at(1).simulate('click');
    expect(inst.onPageChange).toHaveBeenCalled();
  });

  it('should list departments with Pagination set to 1', () => {
    wrapper.setProps({
      location: {
        search: true,
      },
      department: {
        ...props.department,
        isLoadingDepartment: false,
        pagination: {
          pageCount: 2,
          totalCount: 11
        }
      }
    });

    const inst = wrapper.instance();

    inst.onPageChange('previous');
    inst.onPageChange = jest.fn();

    inst.onPageChange('previous');
    wrapper.find('.pagination Button .pagination__button').at(1).simulate('click');
    expect(inst.onPageChange).toHaveBeenCalled();
  });
});
