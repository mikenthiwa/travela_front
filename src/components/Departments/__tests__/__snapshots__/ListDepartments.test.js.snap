// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListDepartments /> renders appropriately 1`] = `
<div
  className="list-reasons"
>
  <div
    className="table__container"
  >
    <table
      className="mdl-data-table mdl-js-data-table readiness-table"
    >
      <thead>
        <tr>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            Departments
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            <span />
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            Parent Department
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            <span />
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            Created By
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            <span />
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          >
            Created On
          </th>
          <th
            className="mdl-data-table__cell--non-numeric table__head"
          />
        </tr>
      </thead>
      <tbody
        className="table__body"
      >
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Operations
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Technology
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Success
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Facilities
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Partner engineering
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Tdd
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Apprenticeship
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Finance
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Travel
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Marketing
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
        <tr
          className="table__rows"
        >
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            Communications
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data readiness__cell-name"
          >
            None
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            BambooHR
          </td>
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          />
          <td
            className="mdl-data-table__cell--non-numeric table__data"
          >
            Invalid date
          </td>
          <td
            className="table__data"
          >
            <span>
              <i
                className="fa fa-ellipsis-v on"
                onClick={[Function]}
                role="presentation"
              />
              <div
                className="context-menu table__menu-container false"
              >
                <ul
                  className="table__menu-list"
                >
                  <li
                    className="edit"
                    onClick={[Function]}
                    role="presentation"
                  >
                    <span>
                      Edit
                    </span>
                  </li>
                </ul>
              </div>
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
